Fix boot in other cases

Rudd-O opened the pull request:
Older distros need these fixes, and there's a race condition fix between pool importers here.  Finally, there's a problem that users can be led into, by setting the supported rootflags= parameters to a value that precludes booting from ZFS.

prometheanfire commented:
I'll add this to my local review queue, at a conf this week so tight on time.

prometheanfire commented:
it looks good as a onceover though

jyxent commented:
I've added a check for grep and made the generator more in line with the dmsquash one here: https://github.com/zfsonlinux/zfs/pull/4562  My boot was failing because grep was missing.

I have a couple comments on these changes:

I don't think there's a race in the import services.  One will run based on whether the cache is there or not.

I don't think 'Before=dracut-mount.service' should be added to the import services, as they will be used on systems that have systemd, but not dracut.

Rudd-O commented:
I would prefer if you did not submit this.

One thing: both the check and the requirement for grep are necessary.  Can you introduce them in a separate PR?  I would approve that right away.

jyxent commented on etc/systemd/system/zfs-import-scan.service.in:
```diff
@@ -4,6 +4,7 @@ DefaultDependencies=no
 After=systemd-udev-settle.service
 After=cryptsetup.target
+After=zfs-import-cache.service
 ConditionPathExists=!@sysconfdir@/zfs/zpool.cache
```
The ConditionPathExists option prevents them from running at the same time.

zfs-import-cache only runs if @sysconfdir@/zfs/zpool.cache exists.
zfs-import-scan only runs if @sysconfdir@/zfs/zpool.cache does not exist.

jyxent replied:
ConditionPathExists prevents the services from running at the same time.

zfs-import-cache only runs if @sysconfdir@/zfs/zpool.cache exists
zfs-import-scan only runs if @sysconfdir@/zfs/zpool.cache doesn't exist

jyxent commented on etc/systemd/system/zfs-import-cache.service.in:
```diff
@@ -4,6 +4,7 @@ DefaultDependencies=no
 Requires=systemd-udev-settle.service
 After=systemd-udev-settle.service
 After=cryptsetup.target
+Before=dracut-mount.service
```
Won't this cause issues with the service for people who have systemd, but not dracut?

Rudd-O replied:
No.  Before=X.unit is ignored if X.unit is nonexistent / cannot be loaded.

jyxent commented on etc/systemd/system/zfs-import-scan.service.in:
```diff
@@ -4,6 +4,8 @@ DefaultDependencies=no
 Requires=systemd-udev-settle.service
 After=systemd-udev-settle.service
 After=cryptsetup.target
+After=zfs-import-cache.service
```
ConditionPathExists prevents the services from running at the same time.

zfs-import-cache only runs if @sysconfdir@/zfs/zpool.cache exists
zfs-import-scan only runs if @sysconfdir@/zfs/zpool.cache doesn't exist

Rudd-O replied:
Good point.  I will eliminate this commit.

jyxent commented on contrib/dracut/90zfs/zfs-generator.sh.in:
```diff
@@ -15,8 +15,10 @@ type getarg >/dev/null 2>&1 || . "$dracutlib"
+    rootflags="zfsutil,${rootflags}"
 else
     rootflags=zfsutil
 fi
```
Not specific to this change, but you could add a few lines here to handle the zfs:AUTO case, where no root was specified, like in mount-zfs.sh.  I assume we don't need to do the check twice, since the pool should already be imported.  Something like this:

```
if [ "${root}" = "zfs:AUTO" ] ; then
  root="$(find_bootfs)"
  if [ $? -ne 0 ] ; then
    exit 0
  fi
fi
```

Rudd-O replied:
I would like that to be the case, but, unfortunately for us, this is far too early in the boot process to be running anything like find_bootfs.  Unlike during the execution of mount-zfs.sh, the pools are not imported at this time, you gotta understand, this is the EARLIEST part of the boot process.  And, after the generators have executed, there is no way to alter the mount unit any further.

So, it looks like there is no way to support that feature (at least not without major rethinking of how it works).  The only way I see to make zfs:AUTO work, would be to require the execution of mount-zfs.sh unconditionally, as well as to require the complete suppression of sysroot.mount — and who knows what other hellish effects that may have!

To be honest, it's not a big deal for people who are using https://github.com/Rudd-O/zfs-fedora-installer/tree/master/grub-zfs-fixer — the root dataset and mountpoint are properly deduced from /etc/fstab in that case, so there is zero need to specify anything on any config file, as the right thing happens upon generation of the grub.cfg file.

I await until brighter minds find a solution to this problem.

kostyaberger replied:
So will it work if I just add my /etc/fstab to dracut-created initramfs? Actually, my datasets are all set to legacy type mountpoint, so I DO have /etc/fstab. 
(It's because I have different OSes on different datasets in the same rpool, that's the whole thing about my system...)

Rudd-O replied:
Nothing about what I said has anything to do with adding fstab to your initrd.  Additionally, nothing in this changeset changes anything about the way that file systems are mounted *after* the system is done with the initial RAM disk.

The program I linked above is simply an enhancement that inspects /etc/fstab *at the time of the creation of grub.cfg*, and then adds the proper `root=ZFS/dataset` parameter to the kernel command line.  This only happens when the system is already fully booted and you (or something else) run `grub2-mkconfig`.

kostyaberger replied:
Sorry, my mistake :). I never use these dracut or grub functionalities and always edit things manually, but now I recall having read about this in the past.

jyxent replied:
Sorry, you're right about the generator script, but I realized there is one way that we can do the auto boot.  If root is unset in the kernel command line and the initramfs, systemd-fstab-generator will not create sysroot.mount.  If zfs-generator bails when it sees "zfs:AUTO", there will no sysroot.mount, so mount-zfs.sh will run and do the auto mount.  Just need this in zfs-generator.sh:

```bash
if [ "${root}" = "zfs:AUTO" ] ; then
  exit 0
fi
```

I just booted using this patch set and that change, and it works with root unspecified.

Rudd-O replied:
I am inclined to believe you, but, have you verified that there will be no sysroot.mount in all supported distributions and versions?  At the very least, a test for this behaviour would include Fedora 20 and 23, and CentOS 7 and 6, verifying that systemctl show sysroot.mount's LoadState is empty (or, alternatively, that they boot).


Where should that snippet go?

jyxent replied:
I've only tested on my Gentoo install (systemd 229 & dracut 044).  I think this is the first commit where systemd-fstab-generator started creating root mounts (https://github.com/systemd/systemd/commit/5e398e546ea65751e6a774daf828fe06f74434a2).  There and in the current code, it doesn't create a mount if there is no root option on the kernel command line.

Dracut may add it's own root option in the initrd, but I use the --no-hostonly-cmdline to strip that off.

That snippet goes just after the if statement that this comment thread belongs to in zfs-generator, just before the two string manipulation commands to remove zfs:/ZFS:.

Rudd-O replied:
Any diff you can come up with, I will add to the patchset if it doesn't interfere with the boot process of anyone else.  Happy to accommodate zfs:AUTO users.

kostyaberger commented:
OK, it boots fine on one of my machines. And since it has the same configuration as do the other 2, I assume it will work on those as well... But I'll test it.

kostyaberger commented:
Doesn't work well on my other, newer machine. Boot stops with sysroot.mount failure -- before ZFS mounting scripts even start, I guess. Because then from the shell I call /usr/lib/dracut/hooks/mount/mount-zfs.sh (it's there) and zpool IS imported and mounted, so after that I resume the boot process.
What I did was installing the RPMs built on the other machine on which I reported above all boots well.

Rudd-O commented:
@kostyaberger need an actual error log.  Please submit error log from systemctl status sysroot.mount so I can begin to understand what the problem is.

Rudd-O commented:
Seems the  failed tests have nothing to do with my patches.

kostyaberger commented:
Not much logging can I report from that failure. The most meaningful part of `systemctl status sysroot.mount is:
filesystem mypool/ROOT/dataset cannot be mounted, unable to open dataset.

But calling /usr/lib/dracut/hooks/mount/*zfs.sh from the emergency shell both imports and mounts the pool all right. 
My guess is sysroot.mount starts on this particular machine before /dev/disk/* appears. I remember having this kind of problem about 1 year ago, then I added this loop at the start of mount-zfs.sh:
```
while [ -z "$(ls /dev/disk/by-id)" ]
do
sleep 3
done
```
and that had fixed it. And the same, the problem existed on one, but not the other machine.

Rudd-O commented:
sysroot.mount is coded to happen *after* the zpool imports.  Perhaps the zpool imports are happening too soon.

systemctl --failed in the dracut shell, discover what other units failed.  systemctl status zpool-import-{scan,cache}.service too.

Rudd-O commented:
Please speed review of this up.  This is preventing people from booting off of ZFS roots.

Also, the failure above has nothing to do with this pull request.  You can check the details link if you so choose.

behlendorf commented:
@prometheanfire @jyxent @kostyaberger can you please post your review / testing comments for this PR.  Including your signed-off-by if you're comfortable with these changes.

@Rudd-O if you can flesh out the commit comments a little more a describe _why_ a chance is need it'll help me review this.   Please rebase this on master and force update the branch while you're at it.

prometheanfire commented:
ya, I'm updating my kernel tonight to 4.4.9, I'll include this test in the update.

prometheanfire commented:
Works for me. :+1: 

However, I'm going to open a bug to try and get it so it only modprobes if it needs to.  Since I build unikernels I have to modify the systemd service files to remove the modprobe line or it will fail.

Signed-off-by: Matthew Thode <mthode@mthode.org>

prometheanfire commented:
Also, as far as the static kernel thing, since afaik, Gentoo is the only one support installs based on that I'll likely just submit a patch to the ebuild to sed out those lines if the `kernel-builtin` use flag is enabled.

@ryao does that sound good?

kostyaberger commented:
No problem. I was just thinking it was worth reporting, but actually I CAN boot. So you can ignore it, because I don't have time right now for a lot of testing.

Sent from Yahoo Mail on Android 
 
  On Thu, May 5, 2016 at 19:02, Rudd-O<notifications@github.com> wrote:   
Please speed review of this up. This is preventing people from booting off of ZFS roots.

Also, the failure above has nothing to do with this pull request. You can check the details link if you so choose.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub

kostyaberger commented:
Ok, it's zfs-import-cache which fails. For the obvious reason there is no zpool.cache file in /etc/zfs or elsewhere. 

Sent from Yahoo Mail on Android 
 
  On Thu, May 5, 2016 at 18:59, Rudd-O<notifications@github.com> wrote:   
sysroot.mount is coded to happen after the zpool imports. Perhaps the zpool imports are happening too soon.

systemctl --failed in the dracut shell, discover what other units failed. systemctl status zpool-import-{scan,cache}.service too.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub

kostyaberger commented:
Well, I deleted the /etc/zfs/zpool.cache to exclude the import by cache service, now it seems there is a bug in /usr/lib/dracut/modules.d/90zfs/mount-zfs.sh:
`loadstate="$(systemctl --system --show -p LoadState sysroot.mount)"`
returns error, because it should be:
`systemctl --system --show -p LoadState | grep sysroot.mount`

kostyaberger commented:
And now that I've fixed the mount-zfs.sh error sysroor.mount fails with this message:filesystem ... cannot be mounted. using 'zfs mount'Use 'zfs set mountpoint=/sysroot' or mount -t zfs ... etc

Sent from Yahoo Mail on Android 
 
  On Fri, May 6, 2016 at 21:22, Kostya Berger<bergerkos@yahoo.co.uk> wrote:   Ok, it's zfs-import-cache which fails. For the obvious reason there is no zpool.cache file in /etc/zfs or elsewhere. 

Sent from Yahoo Mail on Android 
 
  On Thu, May 5, 2016 at 18:59, Rudd-O<notifications@github.com> wrote:   
sysroot.mount is coded to happen after the zpool imports. Perhaps the zpool imports are happening too soon.

systemctl --failed in the dracut shell, discover what other units failed. systemctl status zpool-import-{scan,cache}.service too.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub

kostyaberger commented:
But still, if what I'm talking about has nothing to do with the patches in question, please, make it a separate bug report. Or don't pay attention to it, please.

ryao commented:
@prometheanfire I am fine with using sed on those things when USE=kernel-builtin is specified.

Rudd-O commented:
@kostyaberger we want the systemctl loadstate stuff to return error, not grepping anything at all.  I have added a || true in case this causes some explosion.

@jyxent I have added the "bail if zfs:AUTO" patch to the pull request.

Rudd-O commented:
This works locally.  Please test on your rigs.

We also need to develop a proper automated testing rig.  I could throw some funds to an Amazon account designated to run tests like zfs-fedora-installer image boots (heck, we could use mine own with Amazon IAM), but someone else would need to write the automation to (a) get the program to run (b) boot the image (c) verify that the image booted (d) tear all that shit down when not in use.  Otherwise I would be paying thousands of dollars a month for idle instances, and I really do not want to do that.

Rudd-O commented on contrib/dracut/02zfsexpandknowledge/module-setup.sh.in:
```diff
@@ -17,27 +17,22 @@ get_pool_devices() {
   local resolved
   poolconfigtemp=`mktemp`
-  @sbindir@/zpool list -v -H "$1" > "$poolconfigtemp" 2>&1
+  @sbindir@/zpool list -v -H -P "$1" > "$poolconfigtemp" 2>&1
```
great work @b333z

Rudd-O commented:
Hi folks.  This is preventing people from using ZFS in at least Fedora if not CentOS as well.  If anything, this can be merged as is, and later bugs that may be discovered can be fixed on top of that.  It really should go in.

behlendorf commented:
Merged as:

d402c18 A collection of dracut fixes
